lgc/disassembler/Disassembler.cpp

```diff
@@ -0,0 +1,506 @@
+
+    MCSymbol *sym = m_context->getOrCreateSymbol(symbols.back().Name);
+    switch (symbols.back().Type) {
+    case ELF::STT_FUNC:
+      m_streamer->emitSymbolAttribute(sym, MCSA_ELF_TypeFunction);
+      break;
+    case ELF::STT_OBJECT:
+      m_streamer->emitSymbolAttribute(sym, MCSA_ELF_TypeObject);
+      break;
+    }
+
+    if (uint64_t size = symbolRef.getSize()) {
+      if (symbols.back().Addr + size <= cantFail(sectionRef.getContents()).size()) {
+        synthesizedLabels.push_back(std::make_unique<std::string>((Twine(symbols.back().Name) + "_symend").str()));
+        StringRef endName = *synthesizedLabels.back();
+        MCSymbol *endSym = m_context->getOrCreateSymbol(endName);
+        symbols.push_back(SymbolInfoTy(symbols.back().Addr + size, endName, ELF::STT_NOTYPE));
+        const MCExpr *sizeExpr = MCBinaryExpr::createSub(MCSymbolRefExpr::create(endSym, *m_context),
+                                                         MCSymbolRefExpr::create(sym, *m_context), *m_context);
+        m_streamer->emitELFSize(sym, sizeExpr);
+      }
+    }
+  }
+
+  // Collect and sort the relocs for the section.
+  std::vector<object::RelocationRef> relocs;
+  for (auto &relSect : m_objFile->sections()) {
+    auto relocatedSectOr = relSect.getRelocatedSection();
+    if (relocatedSectOr && *relocatedSectOr == sectionRef) {
+      for (auto &reloc : relSect.relocations())
+        relocs.push_back(reloc);
+    }
+  }
+  std::sort(relocs.begin(), relocs.end(), [](const object::RelocationRef &a, const object::RelocationRef &b) {
+    return a.getOffset() < b.getOffset();
+  });
+
+  // Disassemble the section multiple times until no new synthesized labels appear, then disassemble
+  // one final time actually streaming the output. For non-code, just use a single outputting pass.
+  bool outputting = !isCode;
+  bool isNote = sectType == ELF::SHT_NOTE;
+  for (;;) {
+    // One iteration of disassembling the section.
+    // Sort the symbols. (Stable sort as there may be duplicate addresses.)
+    llvm::stable_sort(symbols);
+
+    // If AMDGPU, create a symbolizer, giving it the symbols.
+    MCSymbolizer *symbolizerPtr = nullptr;
+    if (m_objFile->getArch() == Triple::amdgcn) {
+      std::unique_ptr<MCRelocationInfo> relInfo(m_target->createMCRelocationInfo(m_tripleName, *m_context));
+      if (relInfo) {
+        std::unique_ptr<MCSymbolizer> symbolizer(
+            m_target->createMCSymbolizer(m_tripleName, nullptr, nullptr, &symbols, m_context, std::move(relInfo)));
+        symbolizerPtr = &*symbolizer;
+        m_instDisassembler->setSymbolizer(std::move(symbolizer));
+      }
+    }
+
+    // Get the section contents, and disassemble until nothing left.
+    StringRef contents = cantFail(sectionRef.getContents());
+    size_t offset = 0, lastOffset = 0;
+    auto nextSymbol = symbols.begin();
+    auto nextReloc = relocs.begin();
+
+    for (;;) {
```

Comment:
Could you add a comment explaining when this loop terminates? The loop is quite long and it's hard to immediately tell if it's potentially infinite or not.